src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBioMartRetriever.java

```diff
@@ -0,0 +1,266 @@
+        final int MAX_QUERY_RETRIES = 5;
+        int numLinesProcessed = 0;
+        // Create connection to BioMart URL for each species, retrieving a list of microarray probe types, if available.
+        URL biomartUrlWithSpecies = new URL(queryString);
+        HttpURLConnection biomartConnection = (HttpURLConnection) biomartUrlWithSpecies.openConnection();
+        if (biomartConnection.getResponseCode() == HttpStatus.SC_OK) {
+            Set<String> biomartResponseLines = new HashSet<>();
+            BufferedReader br = new BufferedReader(new InputStreamReader(biomartConnection.getInputStream()));
+            String line;
+            while ((line = br.readLine()) != null) {
+                numLinesProcessed++;
+                // BioMart still responds with a 200, even if no data exists. For now, we handle it by
+                // checking the returned content for the string 'ERROR'. It will retry up to 5 times, with a 10 second delay.
+                if (line.contains("ERROR") ) {
+                    if (retryCount > MAX_QUERY_RETRIES) {
+                        // The data does not exist in BioMart for a few species. Frog (X. tropicalis) has UniProt data but not Microarray data;
+                        // Yeast (S. cerevisiae) has Uniprot-SwissProt data but not UniProt-TrEMBL data; Yeast (S. pombe), P. falciparum
+                        // and D. discoideum don't have UniProt or Microarray data, all at time of writing (January 2020).
+                        throw new BioMartQueryException(line +
+                                "\nThis can happen without issue for certain species (X. tropicalis, S. pombe, S. cerevisiae, P. falciparum) " +
+                                "because the data doesn't exist in BioMart");
+                    }
+
+                    return retryQuery(queryString, retryCount+1, biomartDataType);
+
+                } else {
+                    biomartResponseLines.add(line);
+                }
+
+                if (numLinesProcessed % 10000 == 0) {
+                    logger.info("Processed " + numLinesProcessed + " for " + biomartDataType);
+                }
+            }
+            return biomartResponseLines;
+
+        } else {
+            if (retryCount > MAX_QUERY_RETRIES) {
+                throw new HttpException(
+                        "Unable to connect to BioMart (" +
+                                biomartConnection.getResponseCode() + ": " + biomartConnection.getResponseMessage() +
+                                ") with URL: " + queryString
+                );
+            }
+            return retryQuery(queryString, retryCount+1, biomartDataType);
+        }
+    }
+
+    // Recursive method that retries the BioMart query.
+    private Set<String> retryQuery(String queryString, int retryCount, String biomartDataType) throws InterruptedException, IOException, HttpException, BioMartQueryException {
+        final long QUERY_SLEEP_DURATION = Duration.ofSeconds(5).toMillis();
```

Comment:
Sorry, we might have talked about this but did the duration used to be 10 seconds?

Reply:
Yes, I changed it to 5 seconds though. We've never got any indication that we are querying Ensembl too quickly and since failures are expected (with species that don't have data) I just revised it down

Reply:
Okay, great!